clients/src/main/java/org/apache/kafka/common/security/authenticator/SaslServerAuthenticator.java

```diff
@@ -161,12 +204,69 @@ public void authenticate() throws IOException {
             netInBuffer = null; // reset the networkReceive as we read all the data.
             try {
-                byte[] response = saslServer.evaluateResponse(clientToken);
-                if (response != null) {
-                    netOutBuffer = new NetworkSend(node, ByteBuffer.wrap(response));
-                    flushNetOutBufferAndUpdateInterestOps();
+                switch (saslState) {
+                    case INIT:
+                        String clientMechanism = null;
+                        SaslMechanismResponse mechanismResponse = null;
+                        SaslException exception = null;
+                        try {
+                            SaslMechanismRequest mechanismRequest = new SaslMechanismRequest(ByteBuffer.wrap(clientToken));
+                            clientMechanism = mechanismRequest.mechanism();
+                            if (enabledMechanisms.contains(clientMechanism)) {
+                                LOG.debug("Using SASL mechanism '{}' provided by client", clientMechanism);
+                                mechanismResponse = new SaslMechanismResponse((short) 0, Collections.<String>emptyList());
+                            } else {
+                                LOG.debug("SASL mechanism '{}' provided by client is not supported", clientMechanism);
+                                mechanismResponse = new SaslMechanismResponse(SaslMechanismResponse.UNSUPPORTED_SASL_MECHANISM, enabledMechanisms);
+                                exception = new SaslException("Unsupported SASL mechanism " + clientMechanism);
+                            }
+                            clientToken = null;
+                        } catch (Exception e) {
+                            if (LOG.isDebugEnabled()) {
+                                StringBuilder tokenBuilder = new StringBuilder();
+                                for (byte b : clientToken) {
+                                    tokenBuilder.append(String.format("%02x", b));
+                                    if (tokenBuilder.length() >= 20)
+                                         break;
+                                }
+                                LOG.debug("Received client packet of length {} starting with bytes 0x{}, process as GSSAPI packet", clientToken.length, tokenBuilder);
+                            }
+                            if (enabledMechanisms.contains(SaslConfigs.GSSAPI_MECHANISM)) {
+                                clientMechanism = SaslConfigs.GSSAPI_MECHANISM;
+                                LOG.debug("First client packet is not a SASL mechanism request, using default mechanism GSSAPI");
+                            } else {
+                                exception = new SaslException("Exception handling first SASL packet from client, GSSAPI is not supported by server", e);
+                                mechanismResponse = new SaslMechanismResponse((short) 101, enabledMechanisms);
```

Comment:
@junrao Yes, I wasn't sure whether to send the response in this case. But I thought it would be useful to send it before the connection is closed since it may be useful if you are looking at the bytes returned for debugging purposes.